File: mobile/foodie/lib/models/checkout_session.dart
class CheckoutSession {
  final String id;
  final String status;
  final List<CartItem> cartSnapshot;
  final AddressSnapshot? addressSnapshot;
  final PricingBreakdown pricingBreakdown;
  final AppliedCoupon? appliedCoupon;
  final String paymentMethod;
  final String paymentStatus;
  final DateTime createdAt;

  CheckoutSession({
    required this.id,
    required this.status,
    required this.cartSnapshot,
    this.addressSnapshot,
    required this.pricingBreakdown,
    this.appliedCoupon,
    required this.paymentMethod,
    required this.paymentStatus,
    required this.createdAt,
  });

  factory CheckoutSession.fromJson(Map<String, dynamic> json) {
    return CheckoutSession(
      id: json['_id'] ?? '',
      status: json['status'] ?? 'DRAFT',
      cartSnapshot: (json['cartSnapshot'] as List?)
          ?.map((item) => CartItem.fromJson(item))
          .toList() ?? [],
      addressSnapshot: json['addressSnapshot'] != null
          ? AddressSnapshot.fromJson(json['addressSnapshot'])
          : null,
      pricingBreakdown: PricingBreakdown.fromJson(json['pricingBreakdown'] ?? {}),
      appliedCoupon: json['appliedCoupon'] != null
          ? AppliedCoupon.fromJson(json['appliedCoupon'])
          : null,
      paymentMethod: json['paymentMethod'] ?? 'CASH',
      paymentStatus: json['paymentStatus'] ?? 'UNPAID',
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
    );
  }
}

class CartItem {
  final String dishId;
  final String dishName;
  final String cookId;
  final int quantity;
  final double unitPrice;
  final String? notes;

  CartItem({
    required this.dishId,
    required this.dishName,
    required this.cookId,
    required this.quantity,
    required this.unitPrice,
    this.notes,
  });

  factory CartItem.fromJson(Map<String, dynamic> json) {
    return CartItem(
      dishId: json['dish'] ?? '',
      dishName: json['dishName'] ?? '',
      cookId: json['cook'] ?? '',
      quantity: json['quantity'] ?? 1,
      unitPrice: (json['unitPrice'] ?? 0).toDouble(),
      notes: json['notes'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'dishId': dishId,
      'cookId': cookId,
      'quantity': quantity,
      'unitPrice': unitPrice,
      'notes': notes ?? '',
    };
  }
}

class AddressSnapshot {
  final String fullAddress;
  final String city;
  final String? deliveryNotes;

  AddressSnapshot({
    required this.fullAddress,
    required this.city,
    this.deliveryNotes,
  });

  factory AddressSnapshot.fromJson(Map<String, dynamic> json) {
    return AddressSnapshot(
      fullAddress: json['fullAddress'] ?? '',
      city: json['city'] ?? '',
      deliveryNotes: json['deliveryNotes'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'fullAddress': fullAddress,
      'city': city,
      'coordinates': {'lat': 0, 'lng': 0},
      'deliveryNotes': deliveryNotes ?? '',
    };
  }
}

class PricingBreakdown {
  final double subtotal;
  final double couponDiscount;
  final double autoDiscount;
  final double deliveryFee;
  final double vatAmount;
  final double netTotal;
  final double total;
  final double? vatRate;
  final String? vatLabel;
  final String? currencyCode;
  final String? countryCode;
  final Map<String, dynamic>? debug;

  PricingBreakdown({
    required this.subtotal,
    required this.couponDiscount,
    required this.autoDiscount,
    required this.deliveryFee,
    required this.vatAmount,
    required this.netTotal,
    required this.total,
    this.vatRate,
    this.vatLabel,
    this.currencyCode,
    this.countryCode,
    this.debug,
  });

  factory PricingBreakdown.fromJson(Map<String, dynamic> json) {
    return PricingBreakdown(
      subtotal: (json['subtotal'] ?? 0).toDouble(),
      couponDiscount: (json['couponDiscount'] ?? 0).toDouble(),
      autoDiscount: (json['autoDiscount'] ?? 0).toDouble(),
      deliveryFee: (json['deliveryFee'] ?? 0).toDouble(),
      vatAmount: (json['vatAmount'] ?? 0).toDouble(),
      netTotal: (json['netTotal'] ?? 0).toDouble(),
      total: (json['total'] ?? 0).toDouble(),
      vatRate: (json['vatRate'] ?? 0).toDouble(),
      vatLabel: json['vatLabel'],
      currencyCode: json['currencyCode'],
      countryCode: json['countryCode'],
      debug: json['debug'] is Map<String, dynamic> ? json['debug'] : null,
    );
  }
}

class AppliedCoupon {
  final String code;
  final String campaignId;
  final double discountAmount;

  AppliedCoupon({
    required this.code,
    required this.campaignId,
    required this.discountAmount,
  });

  factory AppliedCoupon.fromJson(Map<String, dynamic> json) {
    return AppliedCoupon(
      code: json['code'] ?? '',
      campaignId: json['campaignId'] ?? '',
      discountAmount: (json['discountAmount'] ?? 0).toDouble(),
    );
  }
}

File: server/models/CheckoutSession.js
const mongoose = require('mongoose');

const checkoutSessionSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  status: {
    type: String,
    enum: ['DRAFT', 'PRICED', 'PAYMENT_PENDING', 'CONFIRMED', 'FAILED', 'EXPIRED'],
    default: 'DRAFT'
  },
  // Cart snapshot
  cartSnapshot: [{
    cook: {
      type: mongoose.Schema.Types.Mixed, // Allow ObjectId or String for demo/legacy data
      ref: 'User'
    },
    dish: {
      type: mongoose.Schema.Types.Mixed, // Allow ObjectId or String for demo/legacy data
      ref: 'Product'
    },
    dishName: String,
    quantity: {
      type: Number,
      required: true
    },
    unitPrice: {
      type: Number,
      required: true
    },
    notes: String,
    fulfillmentMode: {
      type: String,
      enum: ['pickup', 'delivery'],
      default: 'pickup'
    },
    deliveryFee: {
      type: Number,
      default: 0
    },
    prepTime: Number,
    prepReadyConfig: mongoose.Schema.Types.Mixed
  }],
  // Cook preferences (combine/separate timing)
  cookPreferences: {
    type: mongoose.Schema.Types.Mixed,
    default: {}
  },
  // Address snapshot (complete address at checkout time)
  addressSnapshot: {
    addressLine1: String,
    addressLine2: String,
    city: String,
    countryCode: String,
    label: String,
    deliveryNotes: String,
    lat: Number,
    lng: Number
  },
  // Pricing breakdown
  pricingBreakdown: {
    subtotal: { type: Number, default: 0 },
    couponDiscount: { type: Number, default: 0 },
    autoDiscount: { type: Number, default: 0 },
    deliveryFee: { type: Number, default: 0 },
    vatAmount: { type: Number, default: 0 },
    netTotal: { type: Number, default: 0 },
    total: { type: Number, default: 0 },
    vatRate: Number,
    vatLabel: String,
    checkoutVatEnabled: Boolean,
    invoiceVatEnabled: Boolean,
    invoiceVatRate: Number,
    countryCode: String,
    currencyCode: String,
    debug: mongoose.Schema.Types.Mixed
  },
  // Coupon
  appliedCoupon: {
    code: String,
    campaignId: mongoose.Schema.Types.ObjectId,
    discountAmount: Number
  },
  // Payment
  paymentMethod: {
    type: String,
    enum: ['CASH', 'CARD'],
    default: 'CASH'
  },
  paymentStatus: {
    type: String,
    enum: ['UNPAID', 'PAID', 'FAILED'],
    default: 'UNPAID'
  },
  paymentIntentId: String, // Stripe payment intent ID
  // Idempotency
  idempotencyKey: String,
  // Expiration
  expiresAt: {
    type: Date,
    default: () => new Date(Date.now() + 2 * 60 * 60 * 1000) // 2 hours
  }
}, {
  timestamps: true
});

// Indexes
checkoutSessionSchema.index({ user: 1, status: 1 });
checkoutSessionSchema.index({ idempotencyKey: 1 }, { unique: true, sparse: true });
checkoutSessionSchema.index({ expiresAt: 1 }, { expireAfterSeconds: 0 });

module.exports = mongoose.model('CheckoutSession', checkoutSessionSchema);
